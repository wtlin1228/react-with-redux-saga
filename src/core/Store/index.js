import { configureStore } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
import rootReducer from './rootReducer'

export { default as rootReducer } from './rootReducer'
export { default as rootSaga } from './rootSaga'

export const sagaMiddleware = createSagaMiddleware()

export default function configureAppStore(preloadedState) {
  const store = configureStore({
    reducer: rootReducer,
    middleware: [sagaMiddleware],
    preloadedState,
    enhancers: []
  })

  return store
}
