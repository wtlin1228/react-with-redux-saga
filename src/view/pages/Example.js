import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

// utils

// assets

// actions
import {
  exampleActions,
  exampleAsyncActions,
  exampleConstants,
  exampleSelectors
} from 'core/Example'

// components

// self-defined-components

const Example = ({
  exampleData,
  filter,
  fetchExampleAsync,
  setVisibilityFilter
}) => {
  useEffect(() => {
    fetchExampleAsync()
  }, [fetchExampleAsync])

  const {
    visibilityOptions: { SHOW_ALL, SHOW_HANDSOME, SHOW_BEAUTIFUL }
  } = exampleConstants

  return (
    <div>
      <label htmlFor="filter">
        Filter
        <select
          id="filter"
          value={filter}
          onChange={e => setVisibilityFilter({ option: e.target.value })}
          onBlur={e => setVisibilityFilter({ option: e.target.value })}
        >
          <option value={SHOW_ALL}>All</option>
          <option value={SHOW_HANDSOME}>Handsome</option>
          <option value={SHOW_BEAUTIFUL}>Beautiful</option>
        </select>
      </label>

      <ul>
        {exampleData && exampleData.map(item => <li key={item}>{item}</li>)}
      </ul>
    </div>
  )
}

Example.propTypes = {
  exampleData: PropTypes.array.isRequired,
  filter: PropTypes.string.isRequired,
  fetchExampleAsync: PropTypes.func.isRequired,
  setVisibilityFilter: PropTypes.func.isRequired
}

const mapStateToProps = ({ example }) => ({
  exampleData: exampleSelectors.getVisibleData(example),
  filter: example.visibilityFilter
})

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchExampleAsync: exampleAsyncActions.fetchExampleAsync,
      setVisibilityFilter: exampleActions.setVisibilityFilter
    },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(Example)
