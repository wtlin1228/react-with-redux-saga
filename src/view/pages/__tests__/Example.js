import React from 'react'
import { cleanup, render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

import Example from '../Example'

import { exampleConstants, exampleAsyncActions } from 'core/Example'

const mockStore = configureStore([])
const { visibilityOptions } = exampleConstants

describe('Connected React-Redux Component Example', () => {
  let store
  let component

  beforeEach(() => {
    store = mockStore({
      example: {
        visibilityFilter: visibilityOptions.SHOW_ALL,
        data: []
      }
    })

    store.dispatch = jest.fn()

    component = render(
      <Provider store={store}>
        <Example />
      </Provider>
    )
  })

  afterEach(() => {
    cleanup()
  })

  it('should render with given state from Redux store', () => {
    expect(component).toMatchSnapshot()
  })
  it('should dispatch an async action on mounting', () => {
    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(
      exampleAsyncActions.fetchExampleAsync()
    )
  })
})

describe('Connected React-Redux Component Example with mock data', () => {
  let store
  const mockData = [
    'Leo 1',
    'Leo 2',
    'Cat',
    'Leo 3',
    'Una 1',
    'Rabbit',
    'Una 2',
    'Una 3',
    'Dog'
  ]

  beforeEach(() => {
    store = mockStore({
      example: {
        visibilityFilter: visibilityOptions.SHOW_ALL,
        data: mockData
      }
    })
    store.dispatch = jest.fn()
  })

  afterEach(() => {
    cleanup()
  })

  it('should render with given state from Redux store', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Example />
      </Provider>
    )

    const listNode = getByTestId('example-list')
    expect(listNode.childNodes.length).toBe(mockData.length)
  })
})
