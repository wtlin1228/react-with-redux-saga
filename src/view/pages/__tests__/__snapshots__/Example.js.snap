// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected React-Redux Component Example should render with given state from Redux store 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <label
          for="filter"
        >
          Filter
          <select
            id="filter"
          >
            <option
              value="SHOW_ALL"
            >
              All
            </option>
            <option
              value="SHOW_HANDSOME"
            >
              Handsome
            </option>
            <option
              value="SHOW_BEAUTIFUL"
            >
              Beautiful
            </option>
          </select>
        </label>
        <ul
          data-testid="example-list"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <label
        for="filter"
      >
        Filter
        <select
          id="filter"
        >
          <option
            value="SHOW_ALL"
          >
            All
          </option>
          <option
            value="SHOW_HANDSOME"
          >
            Handsome
          </option>
          <option
            value="SHOW_BEAUTIFUL"
          >
            Beautiful
          </option>
        </select>
      </label>
      <ul
        data-testid="example-list"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
